### PR TITLE
[Checkpoint] Add test for fsdp model state saving and loading with/without resharding

### PR DESCRIPTION
**Pull request by wz337**
As title.

https://github.com/pytorch/pytorch/issues/90960

### CODE REVIEW

File: test/distributed/checkpoint/test_fsdp_model_state.py [hunk 1]
```diff
@@ -0,0 +1,102 @@
...
+            state_dict = {
+                "model": model_2.state_dict(),
+            }
+
+            dist_cp.load_state_dict(
+                state_dict=state_dict,
+                storage_reader=dist_cp.FileSystemReader(CHECKPOINT_DIR),
+                planner=DefaultLoadPlanner(
+                    flatten_state_dict=True,
+                    flatten_sharded_tensors=True,
```
Comment (fduwjj) [inline test/distributed/checkpoint/test_fsdp_model_state.py:65]: Isn't this for 2D?

File: test/distributed/checkpoint/test_fsdp_model_state.py [hunk 2]
```diff
@@ -0,0 +1,102 @@
...
+    def _create_new_dist_group(self):
+        # create new fsdp group for resharding
+        fsdp_0 = dist.new_group(ranks=[0, 2])
+        fsdp_1 = dist.new_group(ranks=[1, 3])
+        if dist.get_rank() % 2 == 0:
+            my_fsdp = fsdp_0
+        else:
+            my_fsdp = fsdp_1
+
+        return my_fsdp
```
Comment (fduwjj) [inline test/distributed/checkpoint/test_fsdp_model_state.py:90]: we skip the test when # of GPU is less than 4. What if we use 8 GPUs for test. Shall we not hard code this logic?

**Review by fduwjj — COMMENTED** [review]
Overall lgtm, leave some comments here.

File: test/distributed/checkpoint/test_fsdp_model_state.py (hunk 2 above)
Comment (wz337) [reply test/distributed/checkpoint/test_fsdp_model_state.py]: Thanks! Nice catch! Changed it to form two new dist group (1 for odd number rank and 1 for even number rank).

File: test/distributed/checkpoint/test_fsdp_model_state.py (hunk 1 above)
Comment (wz337) [reply test/distributed/checkpoint/test_fsdp_model_state.py:65]: We actually need this even for FSDP only when using StateDictType.SHARDED_STATE_DICT. 
Before this, we can only use StateDictType.LOCAL_STATE_DICT for testing. For example: https://github.com/pytorch/tau/blob/main/test/spmd/checkpoint/test_pg_planner.py

**Review by fduwjj — APPROVED** [review]
LGTM

**Pull request closed without merging**